pkg/adaptation/plugin.go

```diff
@@ -428,6 +447,74 @@ func (p *plugin) qualifiedName() string {
 	return kind + ":" + idx + "-" + base + pid
 }
 
+// Authenticate handles the plugin's authentication request.
+func (p *plugin) RequestChallenge(ctx context.Context, req *auth.RequestChallengeRequest) (*auth.RequestChallengeResponse, error) {
+	if p.role != nil {
+		log.Errorf(ctx, "rejecting client, multiple authentication requests")
+		p.close()
+		return nil, fmt.Errorf("multiple authentication requests")
+	}
+
+	log.Infof(ctx, "authenticating client with key %q...", string(req.PublicKey))
+	p.role = &role{}
+
+	role, err := p.r.auth.GetRoleForKey(req.PublicKey)
+	if err != nil {
+		log.Errorf(ctx, "rejecting client with unknown key")
+		p.close()
+		return nil, err
+	}
+
+	p.role.role = role
+	impl, err := auth.Get(req.GetAlgorithm())
+	if err != nil {
+		log.Errorf(ctx, "rejecting client, unknown authentication algorithm %q", req.GetAlgorithm())
+	}
+
+	algo, err := impl.NewWithEphemeralKeys()
+	if err != nil {
+		log.Errorf(ctx, "rejecting client, failed to set up authentication: %v", err)
+		p.close()
+		return nil, err
+	}
+
+	p.role.algo = algo
+	chal, key, err := p.role.algo.Challenge([]byte("XXtmpXXX"), req.PublicKey)
```

Comment:
Judging by the name, I'm assuming this will be replaced? Should this be a random seed?

Reply:
Argh... that was already supposed to be replaced by a random seed. Thanks for spotting it !